Can't built on Ubuntu 14.04

ubuntu 14.04 64bit

    cmake --version
    cmake version 2.8.12.2

    ./build.sh

    Executing make
    Linking CXX static library libCoreClrPal.a
    Error running link command: No such file or directory
    make[2]: *** [src/pal/src/libCoreClrPal.a] Error 2
    make[1]: *** [src/pal/src/CMakeFiles/CoreClrPal.dir/all] Error 2
    make: *** [all] Error 2
    Failed to build coreclr components.

On the file system I see directory `binaries/CMake/src/pal/src/CMakeFiles/CoreClrPal.dir`, but don't see the file `all` inside it.

I am able to reproduce this on Arch Linux. The last few lines of the build are:

    Building ASM object src/vm/wks/CMakeFiles/cee_wks.dir/__/amd64/theprestubamd64.S.o
    clang: warning: argument unused during compilation: '-nostdinc++'
    clang: warning: argument unused during compilation: '-nostdinc++'
    [ 20%] Building ASM object src/vm/wks/CMakeFiles/cee_wks.dir/__/amd64/virtualcallstubamd64.S.o
    clang: warning: argument unused during compilation: '-nostdinc++'
    Linking CXX static library libcee_wks.a
    Error running link command: No such file or directory
    src/vm/wks/CMakeFiles/cee_wks.dir/build.make:5596: recipe for target 'src/vm/wks/libcee_wks.a' failed
    make[2]: *** [src/vm/wks/libcee_wks.a] Error 2
    CMakeFiles/Makefile2:56373: recipe for target 'src/vm/wks/CMakeFiles/cee_wks.dir/all' failed
    make[1]: *** [src/vm/wks/CMakeFiles/cee_wks.dir/all] Error 2
    Makefile:107: recipe for target 'all' failed
    make: *** [all] Error 2
    Failed to build coreclr components.

I ran into the same problem, running `build.sh VERBOSE=1` showed that the problem was with `CMAKE_AR` not being set correctly.

```
Linking CXX static library libCoreClrPal.a
cd /home/lukeh/coreclr2/binaries/CMake/src/pal/src && /usr/bin/cmake -P CMakeFiles/CoreClrPal.dir/cmake_clean_target.cmake
cd /home/lukeh/coreclr2/binaries/CMake/src/pal/src && /usr/bin/cmake -E cmake_link_script CMakeFiles/CoreClrPal.dir/link.txt --verbose=1
CMAKE_AR-NOTFOUND cr libCoreClrPal.a  CMakeFiles/CoreClrPal.dir/arch/i386/context.cpp.o CMakeFiles/CoreClrPal.dir/arch/i386/processor.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/file.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/filecrt.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/finite.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/lstr.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/malloc.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/mbstring.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/misc.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/misctls.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/path.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/printf.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/printfcpp.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/silent_printf.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/string.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/stringtls.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/thread.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/wchar.cpp.o CMakeFiles/CoreClrPal.dir/cruntime/wchartls.cpp.o CMakeFiles/CoreClrPal.dir/debug/debug.cpp.o CMakeFiles/CoreClrPal.dir/exception/console.cpp.o CMakeFiles/CoreClrPal.dir/exception/seh.cpp.o CMakeFiles/CoreClrPal.dir/exception/signal.cpp.o CMakeFiles/CoreClrPal.dir/file/directory.cpp.o CMakeFiles/CoreClrPal.dir/file/disk.cpp.o CMakeFiles/CoreClrPal.dir/file/file.cpp.o CMakeFiles/CoreClrPal.dir/file/filetime.cpp.o CMakeFiles/CoreClrPal.dir/file/find.cpp.o CMakeFiles/CoreClrPal.dir/file/path.cpp.o CMakeFiles/CoreClrPal.dir/file/shmfilelockmgr.cpp.o CMakeFiles/CoreClrPal.dir/handlemgr/handleapi.cpp.o CMakeFiles/CoreClrPal.dir/handlemgr/handlemgr.cpp.o CMakeFiles/CoreClrPal.dir/init/pal.cpp.o CMakeFiles/CoreClrPal.dir/init/sxs.cpp.o CMakeFiles/CoreClrPal.dir/loader/module.cpp.o CMakeFiles/CoreClrPal.dir/loader/modulename.cpp.o CMakeFiles/CoreClrPal.dir/locale/unicode.cpp.o CMakeFiles/CoreClrPal.dir/locale/unicode_data.cpp.o CMakeFiles/CoreClrPal.dir/locale/utf8.cpp.o CMakeFiles/CoreClrPal.dir/map/common.cpp.o CMakeFiles/CoreClrPal.dir/map/map.cpp.o CMakeFiles/CoreClrPal.dir/map/virtual.cpp.o CMakeFiles/CoreClrPal.dir/memory/heap.cpp.o CMakeFiles/CoreClrPal.dir/memory/local.cpp.o CMakeFiles/CoreClrPal.dir/misc/dbgmsg.cpp.o CMakeFiles/CoreClrPal.dir/misc/environ.cpp.o CMakeFiles/CoreClrPal.dir/misc/error.cpp.o CMakeFiles/CoreClrPal.dir/misc/fmtmessage.cpp.o CMakeFiles/CoreClrPal.dir/misc/interlock.cpp.o CMakeFiles/CoreClrPal.dir/misc/miscpalapi.cpp.o CMakeFiles/CoreClrPal.dir/misc/strutil.cpp.o CMakeFiles/CoreClrPal.dir/misc/sysinfo.cpp.o CMakeFiles/CoreClrPal.dir/misc/time.cpp.o CMakeFiles/CoreClrPal.dir/misc/utils.cpp.o CMakeFiles/CoreClrPal.dir/misc/version.cpp.o CMakeFiles/CoreClrPal.dir/objmgr/palobjbase.cpp.o CMakeFiles/CoreClrPal.dir/objmgr/shmobject.cpp.o CMakeFiles/CoreClrPal.dir/objmgr/shmobjectmanager.cpp.o CMakeFiles/CoreClrPal.dir/safecrt/makepath_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/memcpy_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/memmove_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/mbusafecrt.c.o CMakeFiles/CoreClrPal.dir/safecrt/safecrt_input_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/safecrt_output_l.c.o CMakeFiles/CoreClrPal.dir/safecrt/safecrt_output_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/safecrt_winput_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/safecrt_woutput_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/splitpath_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/sprintf.c.o CMakeFiles/CoreClrPal.dir/safecrt/sscanf.c.o CMakeFiles/CoreClrPal.dir/safecrt/strcat_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/strcpy_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/strlen_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/strncat_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/strncpy_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/strtok_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/swprintf.c.o CMakeFiles/CoreClrPal.dir/safecrt/vsprintf.c.o CMakeFiles/CoreClrPal.dir/safecrt/vswprint.c.o CMakeFiles/CoreClrPal.dir/safecrt/wcscat_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/wcscpy_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/wcslen_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/wcsncat_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/wcsncpy_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/wcstok_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/wmakepath_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/wsplitpath_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/xtoa_s.c.o CMakeFiles/CoreClrPal.dir/safecrt/xtow_s.c.o CMakeFiles/CoreClrPal.dir/shmemory/shmemory.cpp.o CMakeFiles/CoreClrPal.dir/sync/cs.cpp.o CMakeFiles/CoreClrPal.dir/synchobj/event.cpp.o CMakeFiles/CoreClrPal.dir/synchobj/semaphore.cpp.o CMakeFiles/CoreClrPal.dir/synchobj/mutex.cpp.o CMakeFiles/CoreClrPal.dir/synchmgr/synchcontrollers.cpp.o CMakeFiles/CoreClrPal.dir/synchmgr/synchmanager.cpp.o CMakeFiles/CoreClrPal.dir/synchmgr/wait.cpp.o CMakeFiles/CoreClrPal.dir/thread/process.cpp.o CMakeFiles/CoreClrPal.dir/thread/thread.cpp.o CMakeFiles/CoreClrPal.dir/thread/threadsusp.cpp.o CMakeFiles/CoreClrPal.dir/thread/tls.cpp.o
Error running link command: No such file or directory
make[2]: *** [src/pal/src/libCoreClrPal.a] Error 2
```

I was able to work around this by patching `binaries/CMake/CMakeFiles/2.8.12.2/CMakeCXXCompiler.cmake` to set CMAKE_AR to `/usr/bin/llvm-ar`, but not sure what the correct fix is for this.

--- a/src/pal/tools/gen-buildsys-clang.sh
+++ b/src/pal/tools/gen-buildsys-clang.sh
@@ -24,6 +24,39 @@ else
   buildtype=$2
 fi
 
+# Locate llvm
+# This can be a little complicated, because the common use-case of Ubuntu with
+# llvm-3.5 installed uses a rather unusual llvm installation with the version
+# number postfixed (i.e. llvm-ar-3.5), so we check for that first.
+desired_llvm_version=3.5
+locate_llvm_exec() {
+  if which $1-$desired_llvm_version > /dev/null 2>&1
+  then
+    echo $(which $1-$desired_llvm_version)
+  elif which $1 > /dev/null 2>&1
+  then
+    echo $(which $1)
+  else
+    exit 1
+  fi
+}
+llvm_ar=$(locate_llvm_exec llvm-ar)
+[[ $? -eq 0 ]] || { echo "Unable to locate llvm-ar"; exit 1; }
+llvm_link=$(locate_llvm_exec llvm-link)
+[[ $? -eq 0 ]] || { echo "Unable to locate llvm-link"; exit 1; }
+llvm_nm=$(locate_llvm_exec llvm-nm)
+[[ $? -eq 0 ]] || { echo "Unable to locate llvm-nm"; exit 1; }
+llvm_objdump=$(locate_llvm_exec llvm-objdump)
+[[ $? -eq 0 ]] || { echo "Unable to locate llvm-objdump"; exit 1; }
+llvm_ranlib=$(locate_llvm_exec llvm-ranlib)
+[[ $? -eq 0 ]] || { echo "Unable to locate llvm-ranlib"; exit 1; }
 
-cmake -DCMAKE_USER_MAKE_RULES_OVERRIDE=$1/src/pal/tools/clang-compiler-override.txt -D_CMAKE_TOOLCHAIN_PREFIX=llvm- -D_CMAKE_TOOLCHAIN_SUFFIX=-3.5 -DCMAKE_BUILD_TYPE=$buildtype $1
+cmake -DCMAKE_USER_MAKE_RULES_OVERRIDE=$1/src/pal/tools/clang-compiler-override.txt \
+  -DCMAKE_AR=$llvm_ar \
+  -DCMAKE_LINKER=$llvm_link \
+  -DCMAKE_NM=$llvm_nm \
+  -DCMAKE_OBJDUMP=$llvm_objdump \
+  -DCMAKE_RANLIB=$llvm_ranlib \
+  -DCMAKE_BUILD_TYPE=$buildtype \
+  $1
 

